### PR TITLE
[TOPI] LRN & L2norm Operator

### PR DESCRIPTION
**Pull request by PariksheetPinjari909**
LRN operator for caffe alexnet model to use with onnx frontend

### CODE REVIEW

**Comment by tqchen** [comment]
@sxjscience can you help review this?

**Comment by sxjscience** [comment]
I'll check the formula next week... This week I decide to focus on the MT demo.

**Comment by sxjscience** [comment]
Looks that the functionality can be implemented by stacking multiple reduce/broadcast ops in TOPI.

**Comment by moonat** [comment]
@sxjscience 
For better readability I have split the implementation as per the LRN equation. This way is also correct right?

**Comment by PariksheetPinjari909** [comment]
@sxjscience Using broadcast/reduce I feel it will be difficult since we need to compute for a selected area based on the "local size" parameter, thats why i used reduce_axis to do the sum operation. Can you give a reference how to use broadcast/reduce operator in this scenario?

**Comment by sxjscience** [comment]
Yes, you are correct. Stacking multiple reduce/bcast seems to be infeasible. I’ll check the implementation in detail this week. Currently too busy with other works...

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by PariksheetPinjari909** [comment]
I will upload L2norm operator implementation also along with this pull request, please help to review together.

**Comment by tqchen** [comment]
ping @sxjscience can you take another look

File: topi/python/topi/nn/l2_norm.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# pylint: disable=invalid-name
+"""TVM operator for l2norm"""
+from __future__ import absolute_import
+import tvm
+
+@tvm.target.generic_func
+def l2norm_instance_nchw(data, eps):
```
Comment (sxjscience) [inline topi/python/topi/nn/l2_norm.py]: We should implement L2Norm with arbitrary shape and axis.

File: topi/python/topi/nn/local_response_norm.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    output : tvm.Tensor
+        4-D output with same shape
+    """
+    assert len(data.shape) == 4, "only support 4-dim lrn"
+    b, c, h, w = data.shape
+    assert (size % 2) == 1, "size should be odd number"
+
+    ##Add padding on left & right of size radius first
+    pad_before = [0, int(size/2), 0, 0]
+    pad_after = [0, int(size/2), 0, 0]
```
Comment (sxjscience) [inline topi/python/topi/nn/local_response_norm.py]: use size // 2 instead of int(size/2)

File: topi/python/topi/nn/local_response_norm.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    pad_data = pad(data, pad_before, pad_after, name="pad_data")
+
+    rxk = tvm.reduce_axis((0, size), name='rxk')
+    sqr_sum = tvm.compute((b, c, h, w), lambda i, l, j, k: tvm.sum(
+        pad_data[i, l + rxk, j, k] * pad_data[i, l + rxk, j, k],
+        axis=rxk))
```
Comment (sxjscience) [inline topi/python/topi/nn/local_response_norm.py]: We can first use square and then use topi.sum, for example:
```python
sqr_sum = topi.sum(topi.square(pad_data), axis=1, keepdims=True)
```

File: topi/python/topi/testing/lrn_python.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                        np.power((bias + (alpha * sqr_sum[i, c, j, k] / \
+                                          size)), beta)
+    for i in range(batch):
+        for c in range(channel):
+            for j in range(height):
+                for k in range(weight):
+                    lrn_out[i, c, j, k] = a_np[i, c, j, k] / \
+                                           sqr_sum_up[i, c, j, k]
+
+    return lrn_out
```
Comment (sxjscience) [inline topi/python/topi/testing/lrn_python.py]: We should move the python implementation to `test_topi_lrn.py`.

File: topi/tests/python/test_topi_lrn.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            return
+        print("Running on target: %s" % device)
+        with tvm.target.create(device):
+            s = topi.generic.schedule_lrn(B)
+        ctx = tvm.context(device, 0)
+        a = tvm.nd.array(a_np, ctx)
+        b = tvm.nd.array(np.zeros(get_const_tuple(B.shape), dtype=dtype), ctx)
+        f = tvm.build(s, [A, B], device)
+        f(a, b)
+        np.testing.assert_allclose(b.asnumpy(), b_np, rtol=1e-1)
```
Comment (sxjscience) [inline topi/tests/python/test_topi_lrn.py]: I feel that `rtol=1e-1` is too loose.

File: topi/python/topi/nn/l2_norm.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    """Perform local response normalisation on the data
```
Comment (sxjscience) [inline topi/python/topi/nn/l2_norm.py]: We should write down the formula like in PyTorch (https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/normalization.py#L9-L36) and TensorFlow (https://www.tensorflow.org/api_docs/python/tf/nn/local_response_normalization).

Comment (sxjscience) [inline topi/python/topi/nn/l2_norm.py]: This is L2Norm instead of local response normalization.

File: topi/python/topi/nn/local_response_norm.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
+# pylint: disable=invalid-name
+"""TVM operator for local response norm compute."""
+from __future__ import absolute_import
+import tvm
+from .pad import pad
+
+@tvm.target.generic_func
+def lrn_nchw(data, size, alpha=0.0001, beta=0.75, bias=2):
+    """Perform local response normalisation on the data
```
Comment (sxjscience) [inline topi/python/topi/nn/local_response_norm.py]: We should write down the formula like in PyTorch (https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/normalization.py#L9-L36) and TensorFlow (https://www.tensorflow.org/api_docs/python/tf/nn/local_response_normalization).

File: topi/python/topi/nn/local_response_norm.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    rxk = tvm.reduce_axis((0, size), name='rxk')
+    sqr_sum = tvm.compute((b, c, h, w), lambda i, l, j, k: tvm.sum(
+        pad_data[i, l + rxk, j, k] * pad_data[i, l + rxk, j, k],
+        axis=rxk))
+
+    sqr_sum_up = tvm.compute((b, c, h, w), lambda i, j, k, l: tvm.intrin.power(
+        (bias + (alpha * sqr_sum[i, j, k, l] / size)), beta))
+
+    return tvm.compute(data.shape,
+                       lambda b, c, h, w: data[b, c, h, w] / sqr_sum_up[b, c, h, w])
```
Comment (sxjscience) [inline topi/python/topi/nn/local_response_norm.py]: I think we can reuse the currently implemented binary broadcasting operators like `topi.broadcast_div`.

**Comment by PariksheetPinjari909** [comment]
Review comments are updated in https://github.com/dmlc/tvm/pull/1051
